File: dotfile/handlers/mas.py
"""Mac App Store handler for managing applications."""

import subprocess
from typing import Dict, Any

from ..core.handler import Status, CheckResult
from ..core.utils import command_exists


class MasHandler:
    """Handler for Mac App Store applications."""
    
    def _get_app_id(self, config: Dict[str, Any]) -> str:
        """Get the app ID from config."""
        # App ID is numeric, but we need it as string for command line
        app_id = config.get("mas", "")
        return str(app_id) if app_id else ""
    
    def check(self, config: Dict[str, Any]) -> CheckResult:
        """Check if app is installed via Mac App Store."""
        name = config.get("name", "App")
        app_id = self._get_app_id(config)
        
        if not app_id:
            return CheckResult(
                status=Status.ERROR,
                message=f"{name} has no mas configuration"
            )
        
        if not command_exists("mas"):
            return CheckResult(
                status=Status.ERROR,
                message="mas CLI is not installed"
            )
        
        try:
            # Check if installed
            result = subprocess.run(
                ["mas", "list"],
                capture_output=True,
                text=True,
                check=True
            )
            
            # Check if app ID appears in the list
            for line in result.stdout.strip().split('\n'):
                if line and app_id in line.split()[0]:
                    return CheckResult(
                        status=Status.INSTALLED,
                        message=f"{name} is installed"
                    )
            
            return CheckResult(
                status=Status.NOT_INSTALLED,
                message=f"{name} is not installed"
            )
                
        except Exception as e:
            return CheckResult(
                status=Status.ERROR,
                message=f"Error checking {name}: {e}"
            )
    
    def install(self, config: Dict[str, Any], dry_run: bool = False) -> bool:
        """Install app via Mac App Store."""
        name = config.get("name", "App")
        app_id = self._get_app_id(config)
        
        if not command_exists("mas"):
            print("✗ mas CLI is not installed")
            return False
        
        if dry_run:
            print(f"Would install {name} (mas install {app_id})")
            return True
        
        try:
            # Try to install the app directly
            print(f"  → Installing {name}...")
            install_result = subprocess.run(
                ["mas", "install", app_id],
                capture_output=True,
                text=True
            )
            
            if install_result.returncode == 0:
                return True
            
            # Check if the error indicates the app needs to be purchased
            error_output = install_result.stderr.lower()
            if "purchase" in error_output or "buy" in error_output or "not purchased" in error_output:
                print(f"  → App not purchased, attempting to purchase {name}...")
                
                # Try to purchase the app
                purchase_result = subprocess.run(
                    ["mas", "purchase", app_id],
                    capture_output=True,
                    text=True
                )
                
                if purchase_result.returncode != 0:
                    print(f"✗ Failed to purchase {name}")
                    print(f"  → {purchase_result.stderr}")
                    return False
                
                # Try to install again after purchase
                print(f"  → Installing {name} after purchase...")
                install_result = subprocess.run(
                    ["mas", "install", app_id],
                    capture_output=True,
                    text=True
                )
                
                if install_result.returncode == 0:
                    return True
                else:
                    print(f"✗ Failed to install {name} after purchase")
                    print(f"  → {install_result.stderr}")
                    return False
            else:
                # Some other error occurred
                print(f"✗ Failed to install {name}")
                print(f"  → {install_result.stderr}")
                return False
            
        except subprocess.CalledProcessError as e:
            print(f"✗ Error installing {name}: {e}")
            return False
        except Exception as e:
            print(f"✗ Unexpected error: {e}")
            return False
    
    def uninstall(self, config: Dict[str, Any], dry_run: bool = False) -> bool:
        """Uninstall app via mas uninstall."""
        name = config.get("name", "App")
        app_id = self._get_app_id(config)
        
        if not command_exists("mas"):
            print("✗ mas CLI is not installed")
            return False
        
        if dry_run:
            print(f"Would uninstall {name} (mas uninstall {app_id})")
            return True
        
        try:
            # Uninstall the app
            print(f"  → Uninstalling {name}...")
            result = subprocess.run(
                ["mas", "uninstall", app_id],
                capture_output=True,
                text=True
            )
            
            if result.returncode == 0:
                return True
            else:
                print(f"✗ Failed to uninstall {name}")
                print(f"  → {result.stderr}")
                return False
                
        except subprocess.CalledProcessError as e:
            print(f"✗ Error uninstalling {name}: {e}")
            return False
        except Exception as e:
            print(f"✗ Unexpected error: {e}")
            return False

File: dots/source/handlers/mas.ts
import type { CheckResult, Handler, MacOSConfig, MasConfig } from "../lib/types.js";
import { runCommand, commandExists, appendLog } from "../lib/shell.js";
import { isMas } from "../lib/types.js";

export const masHandler: Handler = {
	async check(config: MacOSConfig, name: string): Promise<CheckResult> {
		if (!isMas(config)) {
			return { status: "error", message: "Invalid mas config" };
		}

		// Check if mas is installed
		if (!(await commandExists("mas"))) {
			return { status: "error", message: "mas CLI is not installed" };
		}

		const masConfig = config as MasConfig;
		const appId = masConfig.mas;

		const result = await runCommand("mas list");
		if (!result.success) {
			return { status: "error", message: "Failed to list installed apps" };
		}

		// Check if app ID is in the list
		const appIdStr = String(appId);
		const lines = result.stdout.split("\n");
		const isInstalled = lines.some((line) => line.startsWith(appIdStr + " "));

		if (isInstalled) {
			return { status: "installed", message: "Installed from App Store" };
		}

		return { status: "not_installed", message: "Not installed" };
	},

	async install(
		config: MacOSConfig,
		name: string,
		dryRun = false,
	): Promise<boolean> {
		if (!isMas(config)) {
			return false;
		}

		const masConfig = config as MasConfig;
		const appId = masConfig.mas;

		if (dryRun) {
			await appendLog(`[DRY-RUN] Would install from App Store: ${appId}`);
			return true;
		}

		await appendLog(`Installing ${name} from App Store (ID: ${appId})`);

		// Try to install
		let result = await runCommand(`mas install ${appId}`);

		// If install fails with "purchase" error, try purchasing first
		if (!result.success && result.stderr.includes("purchase")) {
			await appendLog(`Attempting to purchase ${name} first...`);
			const purchaseResult = await runCommand(`mas purchase ${appId}`);

			if (!purchaseResult.success) {
				await appendLog(`Failed to purchase ${name}: ${purchaseResult.stderr}`);
				return false;
			}

			// Retry install after purchase
			result = await runCommand(`mas install ${appId}`);
		}

		if (!result.success) {
			await appendLog(`Failed to install ${name}: ${result.stderr}`);
			return false;
		}

		await appendLog(`Successfully installed ${name} from App Store`);
		return true;
	},

	async uninstall(
		config: MacOSConfig,
		name: string,
		dryRun = false,
	): Promise<boolean> {
		if (!isMas(config)) {
			return false;
		}

		const masConfig = config as MasConfig;
		const appId = masConfig.mas;

		if (dryRun) {
			await appendLog(`[DRY-RUN] Would uninstall from App Store: ${appId}`);
			return true;
		}

		await appendLog(`Uninstalling ${name} from App Store (ID: ${appId})`);

		const result = await runCommand(`mas uninstall ${appId}`);

		if (!result.success) {
			await appendLog(`Failed to uninstall ${name}: ${result.stderr}`);
			return false;
		}

		await appendLog(`Successfully uninstalled ${name} from App Store`);
		return true;
	},
};
